client/service/src/lib.rs

```diff
@@ -152,26 +152,7 @@ async fn build_network_future<
 	let starting_block = client.info().best_number;
 
 	// Stream of finalized blocks reported by the client.
-	let mut finality_notification_stream = {
-		let mut finality_notification_stream = client.finality_notification_stream().fuse();
-
-		// We tweak the `Stream` in order to merge together multiple items if they happen to be
-		// ready. This way, we only get the latest finalized block.
-		stream::poll_fn(move |cx| {
-			let mut last = None;
-			while let Poll::Ready(Some(item)) =
-				Pin::new(&mut finality_notification_stream).poll_next(cx)
-			{
-				last = Some(item);
-			}
-			if let Some(last) = last {
-				Poll::Ready(Some(last))
-			} else {
-				Poll::Pending
-			}
-		})
-		.fuse()
-	};
+	let mut finality_notification_stream = client.finality_notification_stream().fuse();
```

Comment:
We should keep this behavior in this PR, but this could be a source of bugs at least for justification requests (I actually wonder if it can explain some rare bugs that have been reported). We should probably create an issue to adapt the `on_block_finalized` hooks, to take in the whole notification. This way, e.g. inside sync we could make sure that `extra_justifications.on_block_finalized` gets called for all finalized blocks.

Reply:
The previous behavior was to iterate till the last finalized block and then use that on the `on_block_finalized`.

Because we are now directly receiving the last block I tough that we may get rid of the lookup.

But maybe I don't understood what you are suggesting :-)

Regarding the `on_block_finalized` modification to take the full notification, yes I agree with you that may be a good idea for another issue.